### DISCUSSION TITLE
Where is the documentation?

### DISCUSSION THREAD

**Discussion by drmacchius** [RomanKubiak/ctrlr#589] [General]
Hi World,
I would like to do any plugin editor librariar for my Instruments. Some of there have just a panel, aber I would like to add any function. Any other have no panel.

In the old Internate Page I do not find the Dokumentation. 

Where kann I Find it?

If I build a panel can I automatically port it in MacOS, Linux and Windows or I must chance it to port the panel on an other OS?

**Comment by damiensellier** [comment]
Not sure what you're lookig for as the documentation.
Here is the link to @dobo365 excellent starter guide:
https://github.com/RomanKubiak/ctrlr/files/13347465/Ctrlr.-.Step.by.Step.-.v13.pdf

Panels are working like extensions compatible for any ctrlr ports.  So you can run them on any platform available.

**Reply by drmacchius** [reply]
Thanks very helpfull.
Ist there annegare a complet manual on CTRL? Where can I reed more about Sysex? 
I would like to add a Librarian function to free Panel (for example D50 PG1000 clone). It isn’t possible? 
or Upgrade the Yamaha Fs1r panel and do it like this https://synth-voice.sakura.ne.jp/fs1r_editor_english.html

Is it possible?

**Comment by dobo365** [comment]
Hi, there is no complete manual. You need to plunge, do some first simple panel without fancy stuff then grow from there by experiencing but reading a lot the Ctrlr forum but also the Juce API functions.
You can do a lot without coding but as soon as you will want something more complex you will need coding.
My panels are full of code... Several thousands lines... Even if I'm also using all the available properties and that my code is optimized (all the same type of components are re-using the same functions...).
**With Ctrlr based on Juce and the included Lua coding you can achieve anything you want when you spend the time.**
I have done patch cables drawing, wave manager, librarian...

So, the answer to your question is: yes, it is possible.

**Reply by drmacchius** [reply]
Possible but not easy

**Reply by dobo365** [reply]
As I said, a librarian requires coding and depending on your skills / ambition level will require you several weeks of work.
I usually take those challenging tasks as summer hobby project...

**Reply by drmacchius** [reply]
Is it Not possible copy-Paste the Code vor the function from other Panels?

**Comment by drmacchius** [comment]
A painful ask.

Is it easier with ctrlr to build an editor/librarian plugin for the daw for my vintage instruments or is it easier, faster an with smaller amount of code to do it with NI- Reactor, Max, or Puredata/Plugdata?

**Reply by dobo365** [reply]
Unfortunately, I don't know those. Maybe someone else will be able to answer you.
Vintage instruments can be tricky with their Midi implementation. On the other side, a lot of them had their Midi implementation documented which is less and less the case today.

**Reply by bijlevel** [reply]
First things first. Before you dive into CTRLR, you must make sure you have ALL the information of the MIDI implementation and in particular the sysex implementation/format of your MIDI-device. If that is not available then you have to do an awful lot of reverse engineering and that takes TIME and PATIENCE! If you have a way of getting a sysex dump from your MIDI device, you can use CTRLR to get the sysex and save the data to a file. And then you can start analysing the format. If that's all clear then you can use CTRLR to control your device. I know this advice is rather "abstract', but it's based on my experience. Of course looking at panels/Lua code made by others is also a good way to start and you can indeed copy/paste code from other panels. And also follow @dobo365 's advice: start SIMPLE!

**Reply by drmacchius** [reply]
Bijlevel, thanks for your answer. 
most if my synth have already any ctrlr panel, I like to upgrade the panels or add a librarian function. For any other synth I need just a librarian function in plugin form for the daw/vst-host or they are simple with midi chance to program and do not need a sysex mapping.

My instruments are:
Roland D550
Roland XV 5080
Roland A70e
Roland v-synth gt 
Korg ex8000
korg Grandstage
Novation supernova II
Yamaha fs1r
Waldorf Pulse

For Roland D550, XV5080, Kong EX 8000 there are already panels, but they do not have a librarian function. 

Novantuno Suoerniva II and Waldorf Puose can be totally with midi ch programmed
Roland V-Synth GT can better be programmed on the Instruments than in PC, but I need a librarian Function on Plugin.

roland A70 ex and Körg Grandstage are “stage piano”, they need only a librarian plugin for the daw.

yamaha fs1r use the difficult boy. There is a small panel whiteout librarian. I would like to expand the editor function and copy these free open source editor 
https://synth-voice.sakura.ne.jp/fs1r_editor_english.html
http://www.javelinart.com/fs1r-fseq-editor.html

https://www.gearjunkies.com/2021/01/diy-controller-editor-for-the-yamaha-fs1r/

**Reply by bijlevel** [reply]
Well, all I can add that there is a lot of work for you ahead! It is said before, you can achieve everything with CTRLR, but it takes time and patience. If you have concrete questions, then this forum can try to help you out.
Concerning the FS1R, if the editor is open source, then you should dive into C/C++ to make changes to that editor, I think.
